src/core/size_tracking_channel.h

```diff
@@ -0,0 +1,73 @@
+// Copyright 2023, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#pragma once
+
+#include "core/fibers.h"
+
+namespace dfly {
+
+// SimpleQueue-like interface, but also keeps track over the size of Ts it owns.
+// It has a slightly less efficient TryPush() API as it forces construction of Ts even if they are
+// not pushed.
+// T must have a .size() method, which should return the heap-allocated size of T, excluding
+// anything included in sizeof(T). We could generalize this in the future.
+template <typename T, typename Queue = folly::ProducerConsumerQueue<T>> class SizeTrackingChannel {
+ public:
+  SizeTrackingChannel(size_t n, unsigned num_producers = 1) : queue_(n, num_producers) {
+  }
+
+  // Here and below, we must accept a T instead of building it from variadic args, as we need to
+  // know its size in case it is added.
+  void Push(T t) noexcept {
+    size_ += t.size();
+    queue_.Push(std::move(t));
+  }
```

Comment:
1. For RDB serialization we use multiple snapshots that write into a single channel, so it's not safe to access an non-atomic size
2. If you add this, please also update the code around, so it doesn't become convoluted where there a multiple ways to do the same... For example `RdbSaver::Impl::GetTotalBuffersSize` uses stats for _approximating_ the high bound of channel capacity

Reply:
Also about (2), it will now be the case that because we have two separate memory tracking systems, info memory and memory stats will return different results 🤷🏻‍♂️🤷🏻‍♂️

Reply:
Re/ 1: good comment, will fix, thanks!
Re/ 2: I'm not sure what you mean, if I update `GetTotalBufferSize()` to use the to-be-added `GetTotalChannelCapacity()` method, why would memory stats have different results? Or did you mean something else?

Reply:
No, I just meant that if you don't unify the tracking approaches eventually, they'll not only have duplicated code, but will also diverge in values 🙂

Reply:
Any reason not to modify `GetTotalBufferSize()` to use `GetTotalChannelCapacity()`? Current implementation returns how many total bytes inserted, without ever decreasing that number, it might be hugely over estimating..

Reply:
No, there is no reason not to modify it

PS: It does decrease the number, pulled_bytes grows, so it shouldn't reach zero at the very end 🤔 

```c++
size_t total_bytes = pushed_bytes.load(memory_order_relaxed) + serializer_bytes.load(memory_order_relaxed) - pulled_bytes;
```